README.md
# sdwd
This is the github repository for sdwd package. Below is the CRAN link.
https://cran.r-project.org/web/packages/sdwd/index.html

ChangeLog
CHANGELOG

2015-08-05: version 1.0.2
* Fixed a bug in leave-one-out cross-validation.

2015-01-04: version 1.0.1
* Initial release.
=======================================================================


R/p.sdwd.R
predict.sdwd = function(object, newx, s=NULL, 
    type=c("class", "link"), ...) {
  type = match.arg(type)
  b0 = t(as.matrix(object$b0))
  rownames(b0) = "(Intercept)"
  nbeta = rbind2(b0, object$beta)
  if (!is.null(s)) {
    vnames = dimnames(nbeta)[[1]]
    dimnames(nbeta) = list(NULL, NULL)
    lambda = object$lambda
    lamlist = lambda.interp(lambda, s)
    nbeta = nbeta[ , lamlist$left, drop=FALSE] %*% 
            Diagonal(x=lamlist$frac) +
            nbeta[ , lamlist$right, drop=FALSE] %*% 
            Diagonal(x=1-lamlist$frac)
    dimnames(nbeta) = list(vnames, paste(seq(along=s)))
  }
  nfit = as.matrix(as.matrix(cbind2(1, newx)) %*% nbeta)
  switch(type, link=nfit, class=ifelse(nfit > 0, 1, -1))
} 

print.sdwd = function(x, digits=max(3, 
    getOption("digits")-3), ...) {
  cat("\nCall: ", deparse(x$call), "\n\n")
  print(cbind(Df=x$df, Lambda=signif(x$lambda, digits)))
} 


plot.sdwd = function(x, xvar=c("norm", "lambda"), 
    color=FALSE, label=FALSE, ...) {
  beta = x$beta
  lambda = x$lambda
  df = x$df
  xvar = match.arg(xvar)
  ##beta should be in 'dgCMatrix' format
  which = nonzero(beta)
  beta = as.matrix(beta[which, ])
  xvar = match.arg(xvar)
  switch(xvar, 
    norm = {
      index = colSums(abs(beta))
      iname = "l1 norm"
           }, 
    lambda = {
      index = log(lambda)
      iname = "log lambda"
           }
  )
  xlab = iname
  ylab = "coefficients"
  dotlist = list(...)
  type = dotlist$type
  if (is.null(type)) {
    if (color == FALSE) 
      matplot(index, t(beta), lty=1, xlab=xlab, ylab=ylab, 
        type="l", pch=500, 
        col=gray.colors(12, start=0.05, 
        end=0.7, gamma=2.2), ...) else matplot(index, 
      t(beta), lty=1, xlab=xlab, ylab=ylab, 
      type="l", pch=500, ...)
  } else matplot(index, t(beta), lty=1, xlab=xlab, ylab=ylab, ...)
    atdf = pretty(index)
  prettydf = trunc(approx(x=index, y=df, xout=atdf, rule=2)$y)
  axis(3, at=atdf, labels=prettydf, cex.axis=1, tcl=NA)
  if (label) {
    nnz = length(which)
    xpos = max(index)
    pos = 4
    if (xvar == "lambda") {
      xpos = min(index)
      pos = 2
    }
    xpos = rep(xpos, nnz)
    ypos = beta[, ncol(beta)]
    text(xpos, ypos, paste(which), cex=0.5, pos=pos)
  }
} 


R/cv.sdwd.R
cv.sdwd = function(x, y, lambda=NULL, 
  pred.loss=c("misclass", "loss"), nfolds=5, foldid, ...) {
  ####################################################################
  ## data setup
  if (missing(pred.loss)) 
    pred.loss = "misclass" else pred.loss = match.arg(pred.loss)
  if (!match(pred.loss, c("misclass", "loss"), FALSE)) {
    warning("Only 'misclass' and 'loss' available for 
            DWD; 'misclass' used.")
    pred.loss = "misclass"
  }
  typenames = c(misclass = "mis-classification error",
                loss     = "DWD loss")
  y = c(-1, 1)[as.factor(drop(y))]
  x = as.matrix(x)
  if (!all(y %in% c(-1, 1))) 
    stop("y should be a factor with two levels.")
  x.row = as.integer(NROW(x))
  if (length(y) != x.row) 
    stop("x and y have different number of observations.")  
  ####################################################################
  sdwd.object = sdwd(x, y, lambda=lambda, ...)
  lambda = sdwd.object$lambda
  nz = sapply(coef(sdwd.object, type="nonzero"), length)
  if (missing(foldid)) 
    foldid = sample(rep(seq(nfolds), 
      length=x.row)) else nfolds = max(foldid)
  if (nfolds < 3) 
    stop("nfolds must be bigger than 3; nfolds=5 recommended.")
  outlist = as.list(seq(nfolds))
  ## fit the model nfold times and save them
  for (i in seq(nfolds)) {
    which = foldid == i
    outlist[[i]] = sdwd(x=x[!which, , drop=FALSE], 
                        y=y[!which], lambda=lambda, ...)
  }
  ## select the lambda according to predmat
  fun = paste("cv", class(sdwd.object)[[2]], sep=".")
  cvstuff = do.call(fun, 
    list(outlist, lambda, x, y, foldid, pred.loss))
  cvm = cvstuff$cvm
  cvsd = cvstuff$cvsd
  cvname = cvstuff$name
  out = list(lambda=lambda, cvm=cvm, cvsd=cvsd, 
              cvupper=cvm+cvsd, cvlower=cvm - cvsd, nzero=nz, 
              name=cvname, sdwd.fit=sdwd.object)
  obj = c(out, as.list(getmin(lambda, cvm, cvsd)))
  class(obj) = "cv.sdwd"
  obj
} 

cv.sdwdNET = function(outlist, lambda, x, y, foldid, pred.loss) {
  ### Turn y into c(-1,1)
  y = as.factor(y)
  y = c(-1, 1)[as.numeric(y)]
  nfolds = max(foldid)
  predmat = matrix(NA, length(y), length(lambda))
  nlams = double(nfolds)
  for (i in seq(nfolds)) {
    which = foldid == i
    fitobj = outlist[[i]]
    preds = predict(fitobj, x[which, , drop=FALSE], type="link")
    nlami = length(outlist[[i]]$lambda)
    predmat[which, seq(nlami)] = preds
    nlams[i] = nlami
  }
  typenames = c(misclass = "mis-classification error",
              loss     = "DWD loss")
  cvraw = switch(pred.loss, 
    loss     = dwdloss(y * predmat),
    misclass = (y != ifelse(predmat > 0, 1, -1)))
  if (length(y)/nfolds >= 3) {
    cvob = cvcompute(cvraw, foldid, nlams)
    cvraw = cvob$cvraw
    cvn = cvob$N
  } else cvn = length(y) - colSums(is.na(predmat))    
  cvm = colMeans(cvraw, na.rm=TRUE)
  cvsd = sqrt(colMeans(scale(cvraw, cvm, FALSE)^2, 
    na.rm=TRUE)/(cvn - 1))
  list(cvm = cvm, cvsd = cvsd, name = typenames[pred.loss])
} 

predict.cv.sdwd = function(object, newx, s=c("lambda.1se", 
    "lambda.min"), ...) {
  if (is.numeric(s)) 
    lambda = s else if (is.character(s)) {
      s = match.arg(s)
      lambda = object[[s]]
    } else stop("Invalid form for s")
  predict(object$sdwd.fit, newx, s=lambda, ...)
} 

plot.cv.sdwd = function(x, sign.lambda=1, ...) {
  cvobj = x
  xlab = "log(Lambda)"
  if (sign.lambda < 0) 
    xlab = paste0("-", xlab)
  plot.args = list(x=sign.lambda * log(cvobj$lambda), 
    y=cvobj$cvm, type="n", xlab=xlab, ylab=cvobj$name,
    ylim=range(cvobj$cvupper, cvobj$cvlo))
  new.args = list(...)
  if (length(new.args)) 
    plot.args[names(new.args)] = new.args
  do.call("plot", plot.args)
  error.bars(sign.lambda * log(cvobj$lambda), cvobj$cvupper, 
    cvobj$cvlo, width=0.01, col="darkgrey")
  points(sign.lambda * log(cvobj$lambda), cvobj$cvm, 
    pch=20, col="red")
  axis(side=3, at=sign.lambda * log(cvobj$lambda), 
    labels=paste(cvobj$nz), tick=FALSE, line=0)
  abline(v=sign.lambda * log(cvobj$lambda.min), lty=3)
  abline(v=sign.lambda * log(cvobj$lambda.1se), lty=3)
  invisible()
} 

coef.cv.sdwd = function(object, 
    s=c("lambda.1se", "lambda.min"), ...) {
  if (is.numeric(s)) 
    lambda = s else if (is.character(s)) {
      s = match.arg(s)
      lambda = object[[s]]
    } else stop("Invalid form for s.")
  coef(object$sdwd.fit, s=lambda, ...)
} 


src/sdwd.f90
! --------------------------------------------------------------------------
! sdwd.f90: the algorithm for the sparse distance weighted discrimination. 
! --------------------------------------------------------------------------
!
! USAGE:
! 
! call sdwd (lam2, nobs, nvars, x, y, jd, pf, pf2, dfmax, &
! & pmax, nlam, flmin, ulam, eps, isd, maxit, strong, nalam, b0, beta, &
! & ibeta, nbeta, alam, npass, jerr, strong) 
! 
! INPUT ARGUMENTS:
! 
!    lam2 = regularization parameter for the quadratic penalty of the coefficients
!    nobs = number of observations
!    nvars = number of predictor variables
!    x(nobs, nvars) = matrix of predictors, of dimension N * p; each row is an observation vector.
!    y(nobs) = response variable. This argument should be a two-level factor {-1, 1} 
!            for classification.
!    jd(jd(1)+1) = predictor variable deletion flag
!                  jd(1) = 0  => use all variables
!                  jd(1) != 0 => do not use variables jd(2)...jd(jd(1)+1)
!    pf(nvars) = relative L1 penalties for each predictor variable
!                pf(j) = 0 => jth variable unpenalized
!    pf2(nvars) = relative L2 penalties for each predictor variable
!                pf2(j) = 0 => jth variable unpenalized
!    dfmax = limit the maximum number of variables in the model.
!            (one of the stopping criterion)
!    pmax = limit the maximum number of variables ever to be nonzero. 
!           For example once beta enters the model, no matter how many 
!           times it exits or re-enters model through the path, it will 
!           be counted only once. 
!    nlam = the number of lambda values
!    flmin = user control of lambda values (>=0)
!            flmin < 1.0 => minimum lambda = flmin*(largest lambda value)
!            flmin >= 1.0 => use supplied lambda values (see below)
!    ulam(nlam) = user supplied lambda values (ignored if flmin < 1.0)
!    eps = convergence threshold for coordinate majorization descent. 
!          Each inner coordinate majorization descent loop continues 
!          until the relative change in any coefficient is less than eps.
!    isd = standarization flag:
!          isd = 0 => regression on original predictor variables
!          isd = 1 => regression on standardized predictor variables
!          Note: output solutions always reference original
!                variables locations and scales.
!    maxit = maximum number of outer-loop iterations allowed at fixed lambda value. 
!            (suggested values, maxit = 100000)
!    strong = wheter to adopt the strong rule to accelerate the algorithm.
! 
! OUTPUT:
! 
!    nalam = actual number of lambda values (solutions)
!    b0(nalam) = intercept values for each solution
!    beta(pmax, nalam) = compressed coefficient values for each solution
!    ibeta(pmax) = pointers to compressed coefficients
!    nbeta(nalam) = number of compressed coefficients for each solution
!    alam(nalam) = lambda values corresponding to each solution
!    npass = actual number of passes over the data for all lambda values
!    jerr = error flag:
!           jerr  = 0 => no error
!           jerr > 0 => fatal error - no output returned
!                    jerr < 7777 => memory allocation error
!                    jerr = 7777 => all used predictors have zero variance
!                    jerr = 10000 => maxval(vp) <= 0.0
!           jerr < 0 => non fatal error - partial output:
!                    Solutions for larger lambdas (1:(k-1)) returned.
!                    jerr = -k => convergence for kth lambda value not reached
!                           after maxit (see above) iterations.
!                    jerr = -10000-k => number of non zero coefficients along path
!                           exceeds pmax (see above) at kth lambda value.
! 
! LICENSE: GNU GPL (version 2 or later)
! 
! AUTHORS:
!    Boxiang Wang (boxiang@umn.edu) and Hui Zou (hzou@stat.umn.edu), 
!    School of Statistics, University of Minnesota.
! 
! REFERENCES:
!    Wang, B. and Zou, H. (2015). Sparse Distance Weighted Discrimination, 
!      School of Statistics, University of Minnesota.

      SUBROUTINE sdwdNET (lam2, nobs, nvars, x, y, jd, pf, pf2, & 
      & dfmax, pmax, nlam, flmin, ulam, eps, isd, maxit, strong, &
      & nalam, b0, beta, ibeta, nbeta, alam, npass, jerr) 

         IMPLICIT NONE
!------- arg types ----------------------------------------------
         INTEGER :: nobs, nvars, nlam, nalam, maxit, dfmax, pmax
         INTEGER :: isd, npass, jerr, jd (*), ibeta (pmax), nbeta(nlam)
         DOUBLE PRECISION :: lam2, flmin, eps
         DOUBLE PRECISION :: x (nobs, nvars), y (nobs)
         DOUBLE PRECISION :: pf (nvars), pf2 (nvars)
         DOUBLE PRECISION :: ulam (nlam), alam (nlam)
         DOUBLE PRECISION :: beta (pmax, nlam), b0 (nlam)
         LOGICAL :: strong
         
!------- local declarations -------------------------------------
         INTEGER :: j, l, nk, ierr, ju (nvars), q_int
         DOUBLE PRECISION :: xmean (nvars), xnorm (nvars), maj (nvars)
         
!------- preliminary step ---------------------------------------
         CALL chkvars (nobs, nvars, x, ju)
         IF (jd(1) > 0) ju(jd(2:(jd(1)+1))) = 0
         IF (Maxval (ju) <= 0) THEN
            jerr = 7777
            RETURN
         END IF
         IF (Maxval (pf) <= 0.0D0) THEN
            jerr = 10000
            RETURN
         END IF
         IF (Maxval (pf2) <= 0.0D0) THEN
            jerr = 10000
            RETURN
         END IF
         pf = Max (0.0D0, pf)
         pf2 = Max (0.0D0, pf2)

!------- first standardize the data -----------------------------
         CALL Standard (nobs, nvars, x, ju, isd, xmean, xnorm, maj)
         CALL sdwdNETpath (lam2, maj, nobs, nvars, x, y, ju, pf, &
         & pf2, dfmax, pmax, nlam, flmin, ulam, eps, maxit, strong, &
         & nalam, b0, beta, ibeta, nbeta, alam, npass, jerr)
         IF (jerr > 0) RETURN  ! check error after calling function

!------- organize beta afterward --------------------------------
         DO l = 1, nalam
            nk = nbeta(l)
            IF (isd == 1) THEN
               DO j = 1, nk
                  beta(j, l) = beta(j, l) / xnorm(ibeta(j))
               END DO
            END IF
            b0(l) = b0(l) - Dot_product (beta(1:nk, l), &
           & xmean(ibeta(1:nk)))
         END DO
         RETURN
      END SUBROUTINE sdwdNET

      SUBROUTINE sdwdNETpath (lam2, maj, nobs, nvars, x, y, &
      & ju, pf, pf2, dfmax, pmax, nlam, flmin, ulam, eps, maxit, &
      & strong, nalam, b0, beta, m, nbeta, alam, npass, jerr)
      
         IMPLICIT NONE
!------- arg types -----------------------------------------------
         DOUBLE PRECISION, PARAMETER :: BIG = 9.9E30
         DOUBLE PRECISION, PARAMETER :: MFL = 1.0E-6
         INTEGER, PARAMETER :: MNLAM = 6
         INTEGER :: mnl, nobs, nvars, dfmax, pmax, nlam
         INTEGER :: maxit, nalam, npass, jerr 
         INTEGER :: ju (nvars), m (pmax), nbeta (nlam)
         DOUBLE PRECISION :: lam2, eps
         DOUBLE PRECISION :: x (nobs, nvars), y (nobs)
         DOUBLE PRECISION :: pf (nvars), pf2 (nvars),  ulam (nlam)
         DOUBLE PRECISION :: beta (pmax, nlam), b0 (nlam)
         DOUBLE PRECISION :: alam (nlam), maj (nvars)
         LOGICAL :: strong
         
!------- local declarations -------------------------------------
         DOUBLE PRECISION :: d, dif, oldb, u, v, al, alf, flmin
         DOUBLE PRECISION :: dl (nobs), r (nobs)
         DOUBLE PRECISION :: b (0:nvars), oldbeta (0:nvars)
         INTEGER :: i, k, j, l, ctr, ierr, ni, me, mm (nvars)

!------- local declarations for the strong rule -----------------
         INTEGER :: jx, jxx (nvars)
         DOUBLE PRECISION :: tlam, ga (nvars), vl (nvars), al0
         
!------- some initial setup ------------------------------------- 
         al = 0.0D0
         r = 0.0D0
         b = 0.0D0
         oldbeta = 0.0D0
         m = 0
         mm = 0
         npass = 0
         ni = npass
         maj = maj * 4.0
      
! strong rule
!   jxx = 0 using the strong rule
!   jxx = 1 not using the strong rule
         jxx = 0
         IF (strong .EQV. .TRUE.) THEN
            jxx = 0
         ELSE
            jxx = 1
         END IF
         
!---------- lambda loop -----------------------------------------
         mnl = Min (MNLAM, nlam)
         IF (flmin < 1.0D0) THEN
            flmin = Max (MFL, flmin)
            alf = flmin ** (1.0D0 / (nlam-1.0D0))
         END IF
            vl = 0.0D0
            CALL DWDdrv(nobs, nvars, x, y, r, vl)
            ga = Abs (vl)

         loop_lambda: DO l = 1, nlam
            al0 = al
            IF (flmin >= 1.0D0) THEN
               al = ulam(l)
            ELSE
               IF (l > 2) THEN
                  al = al * alf
               ELSE IF (l == 1) THEN
                  al = BIG
               ELSE IF (l == 2) THEN
                  al0 = 0.0D0
                  DO j = 1, nvars
                  IF (ju(j) /= 0) THEN
                     IF (pf(j) > 0.0D0) THEN
                        al0 = Max (al0, ga(j) / pf(j))
                     END IF
                  END IF
               END DO
               al = al0 * alf
               END IF
            END IF
            ctr = 0
            
!---------- check strong rule (in lamdba loop) ------------------
            tlam = 2.0 * al - al0
            loop_strong_rule: DO j = 1, nvars
               IF (jxx(j) == 1) CYCLE
               IF (ga(j) > pf(j) * tlam) jxx(j) = 1
            END DO loop_strong_rule
            
!---------- outer loop ------------------------------------------
            loop_outer: DO
               oldbeta(0) = b(0)
               IF (ni > 0) oldbeta (m(1:ni)) = b(m(1:ni))
               
!---------- middle loop -----------------------------------------
               loop_middle: DO
                  npass = npass + 1
                  dif = 0.0D0
                  
!---------- 1. update beta_j first ( in middle loop) ------------
                  loop_middle_update_betaj: DO k = 1, nvars
                     IF (jxx(k) == 0) CYCLE
                     IF (ju(k) /= 0) THEN
                        oldb = b(k)
                        u = 0.0D0
! Note that most literature define the margin as u, 
!    but the margin is actually r in the code.
! u: 
! \sum_{i=1}^n V'(r_i)y_i x_{ij}
                        DO i = 1, nobs   
                           IF (r(i) > 0.5) THEN
                              dl(i) = -0.25 / (r(i) * r(i))
                           ELSE
                              dl(i) = -1.0D0
                           END IF
                           u = u + dl(i) * y(i) * x(i, k)
                        END DO 
! u:
! M \tilde{\beta_j} - 1/n * (\sum_{i=1}^n V'(r_i)y_i x_{ij} )
                        u = maj(k) * b(k) - u / nobs
                        v = al * pf(k)
                        v = Abs (u) - v
! The update of the coefficients in Algorithm 1:
                        IF (v > 0.0D0) THEN
                           b(k) = Sign (v, u) / (maj(k) + pf2(k) * lam2)
                        ELSE
                           b(k) = 0.0D0
                        END IF
                        d = b(k) - oldb
                        IF (Abs (d) > 0.0D0) THEN
                           dif = Max (dif, 4.0 * d * d)  !!!
                           r = r + y * x(:, k) * d
                           IF (mm(k) == 0) THEN
                              ni = ni + 1
                              IF (ni > pmax) EXIT
                              mm(k) = ni
                              m(ni) = k !indicate which one is non-zero
                           END IF
                        END IF
                     END IF
                  END DO loop_middle_update_betaj
                  
!---------- 2. update intercept ( in middle loop) ---------------
                  IF (ni > pmax) EXIT
                  d = 0.0D0
                  DO i = 1, nobs
                        IF (r(i) > 0.5) THEN
                           dl(i) = -0.25 / (r(i) * r(i))
                        ELSE
                           dl(i) = -1.0D0
                    END IF
                    d = d + dl(i) * y(i)
                  END DO
                  d = - 0.25 * d / nobs !!!!!!!!!
                  IF (d /= 0.0D0) THEN
! New intercept:
! newbeta0 = oldbeta0 - 1/M * \sum_{i=1}^n V'(r_i) y_i / n                  
                     b(0) = b(0) +  d
                     r = r + y * d
                     dif = Max (dif, 4.0 * d * d) !!!
                  END IF
                  IF (dif < eps) EXIT
                  IF (npass > maxit) THEN
                     jerr = -1
                     RETURN
                  END IF
                  
!---------- inner loop ------------------------------------------
                  loop_inner: DO
                     npass = npass + 1
                     dif = 0.0D0
                     
!---------- 1. update beta_j first ( in inner loop) -------------
                     DO j = 1, ni
                        k = m(j)
                        oldb = b(k)
                        u = 0.0D0
                        DO i = 1, nobs   !!!!!!!!!!!!!!!!!!!!!!
                           IF (r(i) > 0.5) THEN
                              dl(i) = -0.25 / (r(i) * r(i))
                           ELSE
                              dl(i) = -1.0D0
                           END IF
                           u = u + dl(i) * y(i) * x(i, k)
                        END DO !!!!!!!!!!!!!!!!!!!!!!
                        u = maj(k) * b(k) - u / nobs
                        v = al * pf(k)
                        v = Abs (u) - v
                        IF (v > 0.0D0) THEN
                           b(k) = Sign (v, u) / (maj(k) + pf2(k) * lam2)
                        ELSE
                           b(k) = 0.0D0
                        END IF
                        d = b(k) - oldb
                        IF (Abs(d) > 0.0D0) THEN
                           dif = Max (dif, 4.0 * d * d) !!!
                           r = r + y * x(:, k) * d
                        END IF
                     END DO
                     
!---------- 2. update intercept ( in inner loop) ----------------
                     d = 0.0D0
                        DO i = 1, nobs   !!!!!!!!!!!!!!!!!!!!!!
                           IF (r(i) > 0.5) THEN
                              dl(i) = -0.25 / (r(i) * r(i))
                           ELSE
                              dl(i) = -1.0D0
                           END IF
                        d = d + dl(i) * y(i)
                     END DO !!!!!!!!!!!!!!!!!!!!!!
                     d = - 0.25 * d / nobs !!!!!!!!!
                     IF (d /= 0.0D0) THEN
                        b(0) = b(0) + d
                        r = r + y * d
                        dif = Max (dif, 4.0 * d * d) !!!
                     END IF
                     IF (dif < eps) EXIT
                     IF (npass > maxit) THEN
                        jerr = -1
                        RETURN
                     END IF
                  END DO loop_inner
               END DO loop_middle
               IF (ni > pmax) EXIT
               
!---------- final check -----------------------------------------
               jx = 0
               IF (4.0 * (b(0) - oldbeta(0)) ** 2 >= eps) jx = 1
               DO j = 1, ni
                  IF (4.0 * (b(m(j)) - oldbeta(m(j))) ** 2 >= eps) THEN
                     jx = 1
                     EXIT
                  END IF
               END DO
               IF (jx /= 0) CYCLE
!---------- check the KKT conditions of the discarded variables--    
               CALL DWDdrv (nobs, nvars, x, y, r, vl)
               ga = Abs (vl)
               DO j = 1, nvars
                  IF (jxx(j) == 1) CYCLE
                  If (ga(j) > al * pf(j)) THEN
                     jxx(j) = 1
                     jx = 1
                  END IF
               END DO
! jx:
! jx checks if some variable violates the KKT conditions. 
               IF (jx == 1) CYCLE
               EXIT
               ctr = ctr + 1
               IF (ctr > maxit) THEN
                  jerr = - l
                  RETURN
               END IF
            END DO loop_outer
            
!---------- final update variable save results ------------------
            IF (ni > pmax) THEN
               jerr = - 10000 - l
               EXIT
            END IF
            IF (ni > 0) beta(1:ni, l) = b(m(1:ni))
            nbeta(l) = ni
            b0(l) = b(0)
            alam(l) = al
            nalam = l
            IF (l < mnl) CYCLE
            IF (flmin >= 1.0D0) CYCLE
            me = Count (beta(1:ni, l) /= 0.0D0)
            IF (me > dfmax) EXIT
         END DO loop_lambda
         RETURN
      END SUBROUTINE sdwdNETpath
      
      SUBROUTINE DWDdrv (nobs, nvars, x, y, r, vl)
         IMPLICIT NONE
         INTEGER :: nobs, nvars, i
         DOUBLE PRECISION :: x (nobs, nvars), y (nobs)
         DOUBLE PRECISION :: r (nobs), vl (nvars), dl (nobs)
         vl = 0.0
         DO i = 1, nobs
            IF (r(i) > 0.5) THEN
               dl(i) = -0.25 / (r(i) * r(i))
               ELSE
                  dl(i) = -1.0D0
               END IF
         END DO
         vl = Matmul(dl * y, x) / nobs
      END SUBROUTINE DWDdrv
 


R/coef.sdwd.R
coef.sdwdNET = function(object, s=NULL, 
    type=c("coefficients", "nonzero"), ...) {
  type = match.arg(type)
  b0 = t(as.matrix(object$b0))
  rownames(b0) = "(Intercept)"
  nbeta = rbind2(b0, object$beta)
  if (!is.null(s)) {
    vnames = dimnames(nbeta)[[1]]
    dimnames(nbeta) = list(NULL, NULL)
    lambda = object$lambda
    lamlist = lambda.interp(lambda, s)
    nbeta = nbeta[,lamlist$left,drop=FALSE] %*% 
      Diagonal(x=lamlist$frac) +
      nbeta[,lamlist$right,drop=FALSE] %*% 
      Diagonal(x=1-lamlist$frac)
    dimnames(nbeta) = list(vnames, paste(seq(along=s)))
  }
  if (type == "coefficients") 
    return(nbeta)
  if (type == "nonzero") 
    return(nonzero(nbeta[-1, , drop=FALSE], bystep=TRUE))
} 

coef.sdwd = function(object, s=NULL, 
    type=c("coefficients", "nonzero"), ...) 
  NextMethod("coef") 


R/sdwd.R
sdwd = function(x, y, nlambda=100, 
    lambda.factor=ifelse(nobs < nvars, 0.01, 1e-04), 
    lambda=NULL, lambda2=0, pf=rep(1, nvars), 
    pf2=rep(1, nvars), exclude, dfmax=nvars + 1, 
    pmax=min(dfmax * 1.2, nvars), standardize=TRUE, 
    eps=1e-08, maxit=1e+06, strong=T) {
  ####################################################################
  #data setup
  method = "dwd"
  # method = match.arg(method)
  this.call = match.call()
  y = drop(y)
  x = as.matrix(x)
  np = dim(x)
  nobs = as.integer(np[1])
  nvars = as.integer(np[2])
  vnames = colnames(x)
  if (is.null(vnames)) 
    vnames = paste0("V", seq(nvars))
  if (length(y) != nobs) 
    stop("x and y have different number of observations")
  ####################################################################
  #parameter setup
  if (length(pf) != nvars) 
    stop("The size of L1 penalty factor must be same as the number of input variables.")
  if (length(pf2) != nvars) 
    stop("The size of L2 penalty factor must be same as the number of input variables.")
  if (lambda2 < 0) 
    stop("lambda2 must be non-negative")
  maxit = as.integer(maxit)
  lam2 = as.double(lambda2)
  pf = as.double(pf)
  pf2 = as.double(pf2)
  isd = as.integer(standardize)
  eps = as.double(eps)
  dfmax = as.integer(dfmax)
  pmax = as.integer(pmax)
  if (!missing(exclude)) {
    jd = match(exclude, seq(nvars), 0)
    if (!all(jd > 0)) 
      stop("Some excluded variables are out of range.")
    jd = as.integer(c(length(jd), jd))
  } else jd = as.integer(0)
  ####################################################################
  #lambda setup
  nlam = as.integer(nlambda)
  if (is.null(lambda)) {
    if (lambda.factor >= 1) 
      stop("lambda.factor should be less than 1.")
    flmin = as.double(lambda.factor)
    ulam = double(1)
  } else {
    #flmin=1 if user define lambda
    flmin = as.double(1)
    if (any(lambda < 0)) 
      stop("The values of lambda should be non-negative.")
    ulam = as.double(rev(sort(lambda)))
    nlam = as.integer(length(lambda))
  }
  ####################################################################
  fit = switch(method, 
                dwd = sdwdpath(x, y, nlam, flmin, ulam, isd, eps, 
                  dfmax, pmax, jd, pf, pf2, maxit, strong, 
                  lam2, nobs, nvars, vnames), 
                )
  if (is.null(lambda)) 
    fit$lambda = lamfix(fit$lambda)
  fit$call = this.call
  ####################################################################
  class(fit) = c("sdwd", class(fit))
  fit
} 

sdwdpath = function(x, y, nlam, flmin, ulam, isd, eps, dfmax, 
                    pmax, jd, pf, pf2, maxit, strong, lam2, 
                    nobs, nvars, vnames) {
  ####################################################################
  #data setup
  y = as.factor(y)
  y = c(-1, 1)[as.numeric(y)]
  if (!all(y %in% c(-1, 1))) 
    stop("y should be a factor with two levels.")
  ####################################################################
  # call Fortran core
  fit = .Fortran("sdwdNET", lam2, nobs, nvars, 
                as.double(x), as.double(y), jd, pf, pf2, dfmax, 
                pmax, nlam, flmin, ulam, eps, isd, maxit, strong,
                nalam=integer(1), b0=double(nlam), 
                beta=double(pmax * nlam), ibeta=integer(pmax), 
                nbeta=integer(nlam), alam=double(nlam), 
                npass=integer(1), jerr=integer(1))
  #################################################################
  # output
  outlist = getoutput(fit, maxit, pmax, nvars, vnames)
  outlist = c(outlist, list(npasses = fit$npass, jerr = fit$jerr))
  class(outlist) = c("sdwdNET")
  outlist
} 
